graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,372 @@
+import graphql.ExecutionResultImpl;
+import graphql.execution.instrumentation.InstrumentationContext;
+import graphql.execution.instrumentation.InstrumentationState;
+import graphql.execution.instrumentation.SimpleInstrumentation;
+import graphql.execution.instrumentation.SimpleInstrumentationContext;
+import graphql.execution.instrumentation.parameters.InstrumentationExecutionParameters;
+import graphql.execution.instrumentation.parameters.InstrumentationFieldParameters;
+import graphql.schema.*;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+/**
+ * A GraphQL Java Instrumentation that computes a max age for an operation based on @cacheControl
+ * directives. See https://www.apollographql.com/docs/apollo-server/performance/caching/
+ */
+public class CacheControlInstrumentation extends SimpleInstrumentation {
+  private final int defaultMaxAge;
+  private final HintCache hintCache = new HintCache();
+
+  private static final String DIRECTIVE_NAME = "cacheControl";
+  private static final String EXTENSION_KEY = "cache-control";
+
+  public CacheControlInstrumentation() {
+    this(0);
+  }
+
+  public CacheControlInstrumentation(int defaultMaxAge) {
+    this.defaultMaxAge = defaultMaxAge;
+  }
+
+  @Override
+  public InstrumentationState createState() {
+    return new CacheControlState();
+  }
+
+  @Override
+  public InstrumentationContext<ExecutionResult> beginField(
+      InstrumentationFieldParameters parameters) {
+    CacheControlState state = parameters.getInstrumentationState();
+    CacheControlPolicy fieldPolicy = new CacheControlPolicy();
+    boolean inheritMaxAge = false;
+
+    GraphQLUnmodifiedType unwrappedReturnType =
+        GraphQLTypeUtil.unwrapAll(parameters.getExecutionStepInfo().getType());
+
+    // There's no way to set a cacheControl directive on the _entities field or the _Entity union in
+    // SDL.
+    // Instead, we can determine the possible concrete types from the representations arguments and
+    // select the
+    // most restrictive cache policy from those types.
+
+    if (unwrappedReturnType.getName().equals(_Entity.typeName)) {
+      Object representations = parameters.getExecutionStepInfo().getArgument(_Entity.argumentName);
+
+      if (representations instanceof List) {
+        typesFromEntitiesArgument(
+                representations, parameters.getExecutionContext().getGraphQLSchema())
+            .stream()
+            .map(type -> hintCache.getHint((GraphQLCompositeType) type))
+            .filter(Optional::isPresent)
+            .forEach(hint -> fieldPolicy.restrict(hint.get()));
+      }
+    }
+
+    // Cache hint on the return type of this field if it's a composite type
+
+    if (unwrappedReturnType instanceof GraphQLCompositeType) {
+      Optional<Hint> hint = hintCache.getHint((GraphQLCompositeType) unwrappedReturnType);
+
+      if (hint.isPresent()) {
+        fieldPolicy.replace(hint.get());
```

Comment:
This call to replace (vs restrict) is copied from my call, but note that I didn't have any other things writing to fieldPolicy before it, and you did.  I think you're relying on the assumption that _Entity doesn't have a directive (and your comment says that too) but this would be even clearer in the `if` at line 72 was an `else if` instead.